### ISSUE TITLE
Build issue on Visual Studio 08. Link issue on Visual Studio 10.

### ISSUE THREAD

**Issue by vamega** [google/protobuf#1958]
In trying to package the latest release as an Anaconda package I ran into some issues on Windows.

Specifically, a compilation error with Visual Studio 2009, and a linking issue with Visual Studio 2010.
Are these compilers supported by the project?

Errors below:

### Visual Studio 2008:
```
C:\conda\conda-bld\work\protobuf-3.0.0\src\google\protobuf\util\field_mask_util.cc(463) : error C2039: 'at' : is not a member of 'std::map<_Kty,_Ty>'
        with
        [
            _Kty=std::string,
            _Ty=google::protobuf::util::`anonymous-namespace'::FieldMaskTree::Node *
        ]
NMAKE : fatal error U1077: 'C:\PROGRA~2\MICROS~1.0\VC\bin\cl.exe' : return code '0x2'
```

[Full Build log](https://ci.appveyor.com/project/conda-forge/protobuf-feedstock/build/1.0.32/job/y7juv5y13s43rtaa)

### Visual Studio 2010
```
[ 94%] Linking CXX executable tests.exe
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kTimestampMinSeconds" (?kTimestampMinSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kTimestampMaxSeconds" (?kTimestampMaxSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kDurationMinSeconds" (?kDurationMinSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kDurationMaxSeconds" (?kDurationMaxSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
tests.exe : fatal error LNK1169: one or more multiply defined symbols found
LINK failed. with 1169
```

[Full Build log](https://ci.appveyor.com/project/conda-forge/protobuf-feedstock/build/1.0.32/job/0isjyh2nki3ed06w)

Would really appreciate any help fixing this.

[Edited title and text after seeing @jakirkham's note]

**Comment by jakirkham** [comment]
I think Visual Studio 2009 up there is suppose to be Visual Studio 2008.

**Comment by jakirkham** [comment]
Should add this is pretty important for Python. As Python 2.7 uses Visual Studio 2008 and Python 3.4 uses Visual Studio 2010, so not working on these implies that those Pythons are not fully supported. As this wasn't a problem in the betas, it seems like this wasn't intentional, but it would be nice to know either way.

**Comment by freudi74** [comment]
I have the same link errors in visual studio 2013 trying to build protocol buffers 3.0.0 for windows using cmake / nmake as described in the README.

-----debug-----
[ 78%] Linking CXX executable tests.exe
libprotobufd.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kTimestampMinSeconds" (?kTimestampMinSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in time_util_test.cc.obj
libprotobufd.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kTimestampMaxSeconds" (?kTimestampMaxSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in time_util_test.cc.obj
libprotobufd.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kDurationMinSeconds" (?kDurationMinSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in time_util_test.cc.obj
libprotobufd.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kDurationMaxSeconds" (?kDurationMaxSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in time_util_test.cc.obj

-------------release-----------------
[ 60%] Linking CXX executable tests.exe
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kTimestampMinSeconds" (?kTimestampMinSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kTimestampMaxSeconds" (?kTimestampMaxSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kDurationMinSeconds" (?kDurationMinSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
libprotobuf.lib(time_util.cc.obj) : error LNK2005: "public: static __int64 const google::protobuf::util::TimeUtil::kDurationMaxSeconds" (?kDurationMaxSeconds@TimeUtil@util@protobuf@google@@2_JB) already defined in map_test.cc.obj
tests.exe : fatal error LNK1169: one or more multiply defined symbols found

**Comment by freudi74** [comment]
After updating to VS2013 update 5, re-running cmake and nmake, my problem went away.

**Comment by jakirkham** [comment]
So I think this should be easily solved if I'm not mistaken. Wouldn't one merely need to change the [call to `at`]( https://github.com/google/protobuf/blob/v3.0.0/src/google/protobuf/util/field_mask_util.cc#L463 ) to use `operator[]`? If so, is there some reason this hasn't happened?

**Comment by jakirkham** [comment]
Forgot about the linking errors, it turns out commit ( https://github.com/google/protobuf/commit/3b3c8abb9635eb3ea078a821a99c9ef29d66dff7 ) added `kTimestampMinSeconds`, `kTimestampMaxSeconds`, `kDurationMinSeconds`, and `kDurationMaxSeconds` in this [header file]( https://github.com/google/protobuf/blob/v3.0.0/src/google/protobuf/util/internal/constants.h#L52-L62 ). Interestingly this [class]( https://github.com/google/protobuf/blob/v3.0.0/src/google/protobuf/util/time_util.h#L58-L62 ) is defining them as members as well. While I don't think they should conflict, I think the linker may be getting confused and noting them as conflicting. Even though it may be a linker bug, it does seem foolhardy to define them in two places like this. IMHO one set should be picked and the other set dropped.

**Comment by jakirkham** [comment]
Was able to resolve this with 2 patches that fit along the lines above. They can be seen below. Built with VS2008, VS2010, VS2015, XCode 6.1, and devtoolset-2's gcc 4.8.2.

First patch to handle the lack of `std::map::at`.

```diff
diff --git src/google/protobuf/util/field_mask_util.cc src/google/protobuf/util/field_mask_util.cc
index 409010a..5c5e842 100644
--- src/google/protobuf/util/field_mask_util.cc
+++ src/google/protobuf/util/field_mask_util.cc
@@ -460,7 +460,7 @@ void FieldMaskTree::TrimMessage(const Node* node, Message* message) {
       reflection->ClearField(message, field);
     } else {
       if (field->cpp_type() == FieldDescriptor::CPPTYPE_MESSAGE) {
-        Node* child = node->children.at(field->name());
+        Node* child = node->children.find(field->name())->second;
         if (!child->children.empty()) {
           TrimMessage(child, reflection->MutableMessage(message, field));
         }
```

Second patch to handle linking issues.

```diff
diff --git src/google/protobuf/util/time_util.cc src/google/protobuf/util/time_util.cc
index 031d019..9a4a53f 100644
--- src/google/protobuf/util/time_util.cc
+++ src/google/protobuf/util/time_util.cc
@@ -44,6 +44,15 @@ namespace util {
 using google::protobuf::Timestamp;
 using google::protobuf::Duration;
 
+// The min/max Timestamp/Duration values we support.
+//
+// For "0001-01-01T00:00:00Z".
+const int64 TimeUtil::kTimestampMinSeconds = -62135596800LL;
+// For "9999-12-31T23:59:59.999999999Z".
+const int64 TimeUtil::kTimestampMaxSeconds = 253402300799LL;
+const int64 TimeUtil::kDurationMinSeconds = -315576000000LL;
+const int64 TimeUtil::kDurationMaxSeconds = 315576000000LL;
+
 namespace {
 static const int kNanosPerSecond = 1000000000;
 static const int kMicrosPerSecond = 1000000;
@@ -142,13 +151,6 @@ int64 RoundTowardZero(int64 value, int64 divider) {
 }
 }  // namespace
 
-// Actually define these static const integers. Required by C++ standard (but
-// omitting them may still work with some compilers).
-const int64 TimeUtil::kTimestampMinSeconds;
-const int64 TimeUtil::kTimestampMaxSeconds;
-const int64 TimeUtil::kDurationMaxSeconds;
-const int64 TimeUtil::kDurationMinSeconds;
-
 string TimeUtil::ToString(const Timestamp& timestamp) {
   return FormatTime(timestamp.seconds(), timestamp.nanos());
 }
diff --git src/google/protobuf/util/time_util.h src/google/protobuf/util/time_util.h
index 1bac089..74f06e3 100644
--- src/google/protobuf/util/time_util.h
+++ src/google/protobuf/util/time_util.h
@@ -55,11 +55,11 @@ class LIBPROTOBUF_EXPORT TimeUtil {
   // The min/max Timestamp/Duration values we support.
   //
   // For "0001-01-01T00:00:00Z".
-  static const int64 kTimestampMinSeconds = -62135596800LL;
+  static const int64 kTimestampMinSeconds;
   // For "9999-12-31T23:59:59.999999999Z".
-  static const int64 kTimestampMaxSeconds = 253402300799LL;
-  static const int64 kDurationMinSeconds = -315576000000LL;
-  static const int64 kDurationMaxSeconds = 315576000000LL;
+  static const int64 kTimestampMaxSeconds;
+  static const int64 kDurationMinSeconds;
+  static const int64 kDurationMaxSeconds;
 
   // Converts Timestamp to/from RFC 3339 date string format.
   // Generated output will always be Z-normalized and uses 3, 6 or 9
```

**Comment by jakirkham** [comment]
Feel free to see more details in PR ( https://github.com/conda-forge/protobuf-feedstock/pull/13 ).

**Comment by xfxyjwf** [comment]
I believe the issue was already fixed in the latest 3.1.0 release (and also fixed in master). See:
https://github.com/google/protobuf/commit/460e7dd7c47b5a3fc290008317c044fac741916a

**Comment by jakirkham** [comment]
The fix for `std::map::at` seems ok.

Not sure if the `static const` fix provided in that commit is sufficient. Had tried basically the same thing before and it hadn't worked. Really had to delay assignment of those variables until the source file to get anything to work.
